### PR TITLE
WIP(Smart Scripts): Separate entry and guid. For Testing only DO NOT APPROVE

### PR DESCRIPTION
**Pull request by malcrom**
## Changes Proposed:
-  Separate entry and guid into their own fields and do away with -guid values.
-  Please test and inform me of any issues.

## Issues Addressed:
- Current system with -guid values is just awful.
- 
## SOURCE:
The Great One

## Tests Performed:
- Compiles and SQL inserts.
- What I have tested works

## How to Test the Changes:
1. Review SQL and test guids

## Known Issues and TODO List:
Not all -guids have been converted yet. The SQL file with SAI updates will continue to be updated over time.

### CODE REVIEW

**Comment by ZhengPeiRu21** [comment]
I understand of the current system can have difficulty, but I do not know if this is the best approach. This would be inconsistent with all other tables and less compatibility with other project database formats. Have we consider a design of two separate tables, such as done with `creature_addon` and `creature_template_addon`? I think this would be a better design than have every row to have an unused column, and would also easier for converting TrinityCore scripts. This is just an opinion, though; I will not complain if it is decide to use entry and guid columns. ^_^

**Comment by malcrom** [comment]
So using separate tables would make it even harder to code. This way you can see all scripting used by entry and not have to guess if there may be a -guid you have to search for.  This system lets to write the combat AI once and once only and then add lines that are by guid only.

I would like TC to use this system but TC is adverse to using anything I suggest and not because of the code.

**Comment by ZhengPeiRu21** [comment]
> So using separate tables would make it even harder to code. This way you can see all scripting used by entry and not have to guess if there may be a -guid you have to search for. This system lets to write the combat AI once and once only and then add lines that are by guid only.
> 
> I would like TC to use this system but TC is adverse to using anything I suggest and not because of the code.

Okay, I am glad we have think through the options. I do not frequently write complex Smart AI so I am not all knowing these troubles. Thank you for the explaining!

**Comment by malcrom** [comment]
Also ids are staying unique to not interfere with condition system.

**Comment by malcrom** [comment]
I do not suggest anyone do complex SAI. It should be mostly Combat, gossip, quest reward, minor visuals, etc.

**Comment by malcrom** [comment]
My main dislike is scripting a guid and having to redo the combat AI each time. Someone may update the combat AI and forget about the -guids.

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 1]
```diff
@@ -102,11 +102,14 @@ void SmartScript::ProcessEventsFor(SMART_EVENT e, Unit* unit, uint32 var0, uint3
 
         if (eventType == e)
         {
-            ConditionList conds = sConditionMgr->GetConditionsForSmartEvent((*i).entryOrGuid, (*i).event_id, (*i).source_type);
-            ConditionSourceInfo info = ConditionSourceInfo(unit, GetBaseObject(), me ? me->GetVictim() : nullptr);
+
```
Comment (acidmanifesto) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:105]:
```suggestion
```

**Review by acidmanifesto — CHANGES_REQUESTED** [review]

File: src/server/game/AI/SmartScripts/SmartScript.cpp (hunk 1 above)
Comment (malcrom) [reply src/server/game/AI/SmartScripts/SmartScript.cpp]: Fixed sweetness

**Comment by neifion-00000000** [comment]
I like it conceptually, might try when I get home later, does guid override entry or does both guid and entry have to match for the SAI to go off when used?

**Comment by malcrom** [comment]
> I like it conceptually, might try when I get home later, does guid override entry or does both guid and entry have to match for the SAI to go off when used?

So npcs will load lines with guid 0 and guid equal to npcs spawn Id. So you can leave guid =0 for anything all spawns use like combat AI. You add a line with a guid for example if one guid of an entry cast a beam. All lines per entry require unique Id value for condition system to work with it. 

I found a new issue with one npc that has a guid script for the one spawn of him and an entry script for a temp spawn. I'm now looking into a way to load temp spawn scripts.

**Comment by malcrom** [comment]
Any ideas on what to set for guid value to only load if npc has no spawnid, temp spawn not loaded from creature or gameobject tables, would be appreciated.

**Comment by malcrom** [comment]
So I have one entry that has one creature table spawn and one temp spawn for quest. Previously the creature table spawn had a -guid SAI and the temp spawn used the entry SAI. With that system that would be impossible because all spawns load 0 guid lines.

I've tested using -1 guid to only load to temp spawns and am asking for feedback. Below is how the SAI would look for a temp spawn and a creature spawn of the same entry:

```
-- Engineer "Spark" Overgrind
DELETE FROM `smart_scripts` WHERE `entry` IN (-61966) AND `source_type`=0;
DELETE FROM `smart_scripts` WHERE `entry` IN (17243) AND `source_type`=0;
INSERT INTO `smart_scripts` (`entry`,`guid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`event_param5`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_param4`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(17243,61966,0,0,1,62,0,100,512,7426,0,0,0,0,80,1724300,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Engineer "Spark" Overgrind - On Gossip Option 0 Selected - Run Script'),
(17243,61966,0,1,0,61,0,100,512,0,0,0,0,0,22,2,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Engineer "Spark" Overgrind - On Gossip Option 0 Selected - Set Event Phase 2'),
(17243,61966,0,2,0,0,0,100,0,0,250,8000,15000,0,11,7978,0,0,0,0,0,2,0,0,0,0,0,0,0,0,'Engineer "Spark" Overgrind - In Combat - Cast ''Throw Dynamite'''),
(17243,61966,0,3,4,1,1,100,0,120000,150000,0,0,0,1,11,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Engineer "Spark" Overgrind - Out of Combat - Say Line 11 (Phase 1)'),
(17243,61966,0,4,0,61,1,100,0,0,0,0,0,0,1,12,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Engineer "Spark" Overgrind - Out of Combat - Say Line 12 (Phase 1)'),
(17243,61966,0,5,0,25,0,100,512,0,0,0,0,0,22,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Engineer "Spark" Overgrind - On Reset - Set Event Phase 1'),
(17243,61966,0,6,7,1,1,100,512,15000,15000,0,0,0,2,875,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Engineer "Spark" Overgrind - Out of Combat - Set Faction 875 (Phase 1)'),
(17243,61966,0,7,0,61,1,100,512,0,0,0,0,0,82,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Engineer "Spark" Overgrind - Out of Combat - Add Npc Flags Gossip (Phase 1)'),
(17243,-1,0,8,9,38,0,100,0,1,1,0,0,0,69,1,0,0,0,0,0,8,0,0,0,0,-5080.7,-11253.6,0.56,0,'Engineer "Spark" Overgrind - On Data Set 1 1 - Move To Position'),
(17243,-1,0,9,0,61,0,100,512,0,0,0,0,0,22,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Engineer "Spark" Overgrind - On Data Set 1 1 - Set Event Phase 1'),
(17243,-1,0,10,0,34,1,100,512,0,1,0,0,0,80,1724301,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Engineer "Spark" Overgrind - On Reached Point 1 - Run Script (Phase 1)'),
(17243,-1,0,11,12,34,1,100,512,0,2,0,0,0,41,250,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Engineer "Spark" Overgrind - On Reached Point 2 - Despawn In 250 ms (Phase 1)'),
(17243,-1,0,12,0,61,1,100,512,0,0,0,0,0,33,17243,0,0,0,0,0,201,30298,0,150,0,0,0,0,0,'Engineer "Spark" Overgrind - On Reached Point 2 - Quest Credit ''NULL'' (Phase 1)');
```

**Comment by malcrom** [comment]
So TC may be doing this a bit differently and I have no idea what their plans are due to that it was an IRC discussion with no future looking ideas posted to be hashed out yet. I'm putting this on hold for now because I do not want to convert 2000 lines of -guid values to find out we are going to mirror TC and possibly have to do it again.

**Comment by temperrr** [comment]
Curiosity, whats wrong with our current system?

**Pull request closed without merging**